### PR TITLE
PLAT-808-3 Create Basic Maintenance Page & Logic

### PR DESCRIPTION
**Pull request by daniel-klose**
- This PR only aims to build a basic management page & basic action for managing maintenances
  - This does not include all the session invalidation, user info and login prevention logic
  - The above mentioned will be done in the next PR

### CODE REVIEW

File: platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/action/StartMaintenanceAction.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import com.softicar.platform.emf.predicate.IEmfPredicate;
+
+public class StartMaintenanceAction implements IEmfManagementAction<AGMaintenance> {
+
+	@Override
+	public IEmfPredicate<AGMaintenance> getPrecondition() {
+
+		return MaintenancePredicates.PENDING//
+			.and(MaintenancePredicates.NEXT_IN_LINE)
+			.and(MaintenancePredicates.NO_MAINTENANCE_ACTIVE);
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/action/StartMaintenanceAction.java:23]: - IMO `NEXT_IN_LINE` und `NO_MAINTENANCE_ACTIVE` should not be added here
- instead, NEXT_IN_LINE should trigger a confirmation in `handleClick`, that they are about to start a maintenance window which is not next in line
- NO_MAINTENANCE_ACTIVE should trigger an exception in `handleClick`

File: platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/status/AGMaintenanceStatus.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package com.softicar.platform.core.module.maintenance.status;
+
+import com.softicar.platform.common.core.entity.IEntity;
+import com.softicar.platform.common.core.exceptions.SofticarUnknownEnumConstantException;
+import com.softicar.platform.common.core.i18n.IDisplayString;
+import com.softicar.platform.core.module.CoreI18n;
+
+public class AGMaintenanceStatus extends AGMaintenanceStatusGenerated implements IEntity {
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/status/AGMaintenanceStatus.java:8]: as already suggested: I think the name `AGMaintenanceWindowState` would be preferable

**Review by oliver-richers — CHANGES_REQUESTED** [review]

File: platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/AGMaintenanceTable.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+	public AGMaintenanceTable(IDbObjectTableBuilder<AGMaintenance> builder) {
+
+		super(builder);
+	}
+
+	@Override
+	public void customizeAuthorizer(EmfAuthorizer<AGMaintenance, SystemModuleInstance> authorizer) {
+
+		authorizer.setCreationRole(CoreRoles.MAINTENANCE_ADMINISTRATOR);
```
[the remark this answers is not in the document]
Comment (daniel-klose) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/AGMaintenanceTable.java:30]: hm oh right, seems i forgot to change my notes

File: platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/AGMaintenanceTable.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		authorizer.setEditRole(CoreRoles.MAINTENANCE_ADMINISTRATOR.toOtherEntityRole());
+		authorizer.setViewRole(CoreRoles.SUPER_USER.toOtherEntityRole());
+	}
+
+	@Override
+	public void customizeEmfTableConfiguration(EmfTableConfiguration<AGMaintenance, Integer, SystemModuleInstance> configuration) {
+
+		configuration.setEditPredicate(MaintenancePredicates.PENDING);
```
[the remark this answers is not in the document]
Comment (daniel-klose) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/AGMaintenanceTable.java:38]: well if you say so

File: platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/MaintenancePredicates.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+		AGMaintenance::isInProgress);
+	IEmfPredicate<AGMaintenance> NEXT_IN_LINE = new EmfPredicate<>(//
+		CoreI18n.NEXT_IN_LINE,
+		AGMaintenance::isNextInLine);
+
+	IEmfPredicate<AGMaintenance> MAINTENANCE_ACTIVE = new EmfPredicate<>(//
+		CoreI18n.MAINTENANCE_ACTIVE,
+		it -> AGMaintenance.isMaintenanceActive());
+
+	IEmfPredicate<AGMaintenance> NO_MAINTENANCE_ACTIVE = MAINTENANCE_ACTIVE.not();
```
[the remark this answers is not in the document]
Comment (daniel-klose) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/MaintenancePredicates.java:24]: well yeah, predicates are scoped to the object, or is there a way to make a predicate without one?

File: platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/AGMaintenance.sqml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+PACKAGE com.softicar.platform.core.module.maintenance
+
+IMPORT com.softicar.platform.core.module.maintenance.status.AGMaintenanceStatus
+
+TABLE AGMaintenance "Core.Maintenance" {
+	PK Integer id = SERIAL
+	DayTime expectedStart
```
[the remark this answers is not in the document]
Comment (daniel-klose) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/AGMaintenance.sqml:7]: IMO this is the way to announce maintenances (as ive seen in games etc)
they always say that the maintenance starts at XX and they expect it to end in XX hours or at XX:XX
Also, as we talked, maintenances can be started early, finished early or later, so expected kinda signifies "around that time" for the users

File: platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/AGMaintenance.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+	public boolean isNextInLine() {
+
+		return AGMaintenance.TABLE//
+			.createSelect()
+			.where(AGMaintenance.STATUS.equal(AGMaintenanceStatusEnum.PENDING.getRecord()))
+			.orderBy(AGMaintenance.EXPECTED_START)
+			.getFirstAsOptional()
+			.map(it -> it.equals(this))
+			.orElse(isPending());
```
[the remark this answers is not in the document]
Comment (daniel-klose) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/AGMaintenance.java:34]: woops, forgot to change this one

File: platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/AGMaintenance.sqml (hunk 6 above)
[the remark this answers is not in the document]
Comment (oliver-richers) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/AGMaintenance.sqml]: ok

File: platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java [hunk 8]
```diff
@@ -8,6 +8,8 @@
 
 public interface CoreI18n extends EmfI18n {
 
+	I18n0 A_MAINTENANCE_IS_ALREADY_IN_PROGRESS = new I18n0("A maintenance is already in progress.")//
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/CoreI18n.java:11]: should be `maintenance window`

File: platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/AGMaintenanceWindow.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+		return AGMaintenanceWindow.TABLE//
+			.createSelect()
+			.where(AGMaintenanceWindow.STATE.equal(AGMaintenanceStateEnum.PENDING.getRecord()))
+			.orderBy(AGMaintenanceWindow.EXPECTED_START)
+			.getFirstAsOptional()
+			.map(it -> it.equals(this))
+			.orElse(false);
+	}
+
+	public static boolean isMaintenanceInProgress() {
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/AGMaintenanceWindow.java:37]: from the name, it is not very obvious that this is a static method, a caller might easily confuse it with the method isInProgress()... suggested name: existsMaintenanceWindowInProgress

File: platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/action/CancelMaintenanceAction.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+import com.softicar.platform.core.module.CoreRoles;
+import com.softicar.platform.core.module.maintenance.AGMaintenanceWindow;
+import com.softicar.platform.core.module.maintenance.MaintenancePredicates;
+import com.softicar.platform.core.module.maintenance.state.AGMaintenanceStateEnum;
+import com.softicar.platform.db.core.transaction.DbTransaction;
+import com.softicar.platform.emf.action.IEmfManagementAction;
+import com.softicar.platform.emf.authorization.role.IEmfRole;
+import com.softicar.platform.emf.predicate.IEmfPredicate;
+
+public class CancelMaintenanceAction implements IEmfManagementAction<AGMaintenanceWindow> {
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/action/CancelMaintenanceAction.java:16]: CancelMaintenanceAction -> CancelMaintenanceWindowAction

File: platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/action/FinishMaintenanceAction.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+import com.softicar.platform.core.module.CoreRoles;
+import com.softicar.platform.core.module.maintenance.AGMaintenanceWindow;
+import com.softicar.platform.core.module.maintenance.MaintenancePredicates;
+import com.softicar.platform.core.module.maintenance.state.AGMaintenanceStateEnum;
+import com.softicar.platform.db.core.transaction.DbTransaction;
+import com.softicar.platform.emf.action.IEmfManagementAction;
+import com.softicar.platform.emf.authorization.role.IEmfRole;
+import com.softicar.platform.emf.predicate.IEmfPredicate;
+
+public class FinishMaintenanceAction implements IEmfManagementAction<AGMaintenanceWindow> {
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/action/FinishMaintenanceAction.java:16]: FinishMaintenanceAction -> FinishMaintenanceWindowAction

File: platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/action/StartMaintenanceAction.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+import com.softicar.platform.core.module.maintenance.MaintenancePredicates;
+import com.softicar.platform.core.module.maintenance.state.AGMaintenanceStateEnum;
+import com.softicar.platform.db.core.transaction.DbTransaction;
+import com.softicar.platform.emf.action.IEmfManagementAction;
+import com.softicar.platform.emf.authorization.role.IEmfRole;
+import com.softicar.platform.emf.predicate.IEmfPredicate;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+public class StartMaintenanceAction implements IEmfManagementAction<AGMaintenanceWindow> {
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/action/StartMaintenanceAction.java:19]: StartMaintenanceAction -> StartMaintenanceWindowAction

File: platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/state/AGMaintenanceState.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
+package com.softicar.platform.core.module.maintenance.state;
+
+import com.softicar.platform.common.core.entity.IEntity;
+import com.softicar.platform.common.core.exceptions.SofticarUnknownEnumConstantException;
+import com.softicar.platform.common.core.i18n.IDisplayString;
+import com.softicar.platform.core.module.CoreI18n;
+
+public class AGMaintenanceState extends AGMaintenanceStateGenerated implements IEntity {
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/maintenance/state/AGMaintenanceState.java:8]: AGMaintenanceState -> AGMaintenanceWindowState

**Review by oliver-richers — APPROVED** [review]
acceptable, but added suggestions

**Review by oliver-richers — APPROVED** [review]

**Pull request merged**